File: test/host_test_webcam.py
import socket
import numpy as np
import cv2
import struct

# Server configuration
HOST = "10.0.0.229"  # Listen on all interfaces
PORT = 5000       # Must match the client's port

# Create and bind socket
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.bind((HOST, PORT))
server_socket.listen(1)
print(f"Listening on {HOST}:{PORT}...")

conn, addr = server_socket.accept()
print(f"Connected by {addr}")

while True:
    # Receive image size (first 4 bytes)
    data = conn.recv(4)
    if not data:
        break
    
    img_size = struct.unpack(">I", data)[0]  # Convert from bytes to int
    print(f"Receiving {img_size} bytes...")

    # Receive image data
    img_data = b""
    while len(img_data) < img_size:
        packet = conn.recv(img_size - len(img_data))
        if not packet:
            break
        img_data += packet

    # Convert received bytes to numpy array
    img_array = np.frombuffer(img_data, dtype=np.uint8)
    img = cv2.imdecode(img_array, cv2.IMREAD_COLOR)

    if img is not None:
        cv2.imshow("Received Image", img)
        cv2.waitKey(0)  # Wait for a key press
        cv2.destroyAllWindows()  # Close the window
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
    else:
        print("Failed to decode image")

conn.close()
server_socket.close()
cv2.destroyAllWindows()

File: test/host_test_webcam.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <winsock2.h>

#pragma comment(lib, "ws2_32.lib")  // Link with ws2_32.lib

#define PORT 5000

int main() {
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);

    SOCKET serverSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (serverSocket == INVALID_SOCKET) {
        std::cerr << "Socket creation failed!" << std::endl;
        return 1;
    }

    sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(PORT);
    serverAddr.sin_addr.s_addr = INADDR_ANY;

    if (bind(serverSocket, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) == SOCKET_ERROR) {
        std::cerr << "Binding failed!" << std::endl;
        return 1;
    }

    listen(serverSocket, 1);
    std::cout << "Server listening on port " << PORT << std::endl;

    SOCKET clientSocket = accept(serverSocket, NULL, NULL);
    if (clientSocket == INVALID_SOCKET) {
        std::cerr << "Failed to accept client connection!" << std::endl;
        return 1;
    }

    std::cout << "Client connected, receiving image..." << std::endl;

    int imgSize;
    recv(clientSocket, (char*)&imgSize, sizeof(imgSize), 0);  // Receive image size

    std::vector<uchar> buffer(imgSize);
    recv(clientSocket, (char*)buffer.data(), imgSize, 0);  // Receive image data

    cv::Mat img = cv::imdecode(buffer, cv::IMREAD_COLOR);  // Convert back to image
    if (img.empty()) {
        std::cerr << "Error: Received empty image!" << std::endl;
        return 1;
    }

    cv::imshow("Received Image", img);
    cv::waitKey(0);

    closesocket(clientSocket);
    closesocket(serverSocket);
    WSACleanup();
    return 0;
}
